### PR TITLE
Extract wheel encoder to separate thread for faster sampling

### LINKED ISSUE #108: Extract wheel encoder to separate thread for faster sampling
Similar to last year's [`process_encoder.py`](https://github.com/UCI-HyperXite/hx8-control-system/blob/main/pod-control/src/components/process_encoder.py), the wheel encoder will not be sampled at a sufficient speed through the normal FSM tick: sensor measurements from I2C will take more than 10 ms whereas the wheel encoder needs to be sampled at around 1.5 ms for the pod's top speed of about 20 mph.
- Explore options for multiprocessing which won't be limited to a single process
  - [Tokio's multi-thread scheduler](https://docs.rs/tokio/latest/tokio/runtime/index.html#multi-thread-scheduler) might be sufficient on its own to have the wheel encoder task run on a dedicated process
- Instead of measuring the wheel encoder in `running_periodic` spawn a new task in `StateMachine::run` which runs its own loop to sample the wheel encoder at a faster rate than the standard FSM tick

### PR DESCRIPTION
**Pull request by ryescholin**
Create separate task thread for sampling wheel encoder. Resolves #108.

### CODE REVIEW

**Review by vrushang1234 — APPROVED** [review]
Looks good

File: pod-operation/src/state_machine.rs [hunk 1]
```diff
@@ -109,18 +109,43 @@ impl StateMachine {
...
+		tokio::spawn(Self::wheel_encoder_task(encoder));
 
 		loop {
 			self.tick().await;
 			interval.tick().await;
 		}
 	}
 
+	async fn wheel_encoder_task(wheel_encoder: std::sync::Arc<std::sync::Mutex<WheelEncoder>>) {
+		let mut interval = tokio::time::interval(ENCODER_SAMPLE_INTERVAL);
```
Comment (taesungh) [inline pod-operation/src/state_machine.rs:129]: Suggestion: could we move this function to be an associated function of `WheelEncoder` and make the sample interval a parameter?

File: pod-operation/src/state_machine.rs [hunk 2]
```diff
@@ -109,18 +109,43 @@ impl StateMachine {
...
+		loop {
+			// Lock the mutex, measure, then immediately unlock
+			let _value = {
+				let mut encoder = wheel_encoder.lock().unwrap();
+				match encoder.measure() {
+					Ok(value) => value,
+					Err(e) => {
+						info!("Wheel encoder error: {:?}", e);
+						continue;
+					}
```
Comment (taesungh) [inline pod-operation/src/state_machine.rs:140]: Note: we'll want to handle the undersampling case, but let's leave that for a separate issue.

File: pod-operation/src/state_machine.rs [hunk 3]
```diff
@@ -226,21 +248,36 @@ impl StateMachine {
...
 
 	/// Perform operations when the pod is running
 	fn _running_periodic(&mut self) -> State {
 		info!("Rolling Running state");
-		let encoder_value = self.wheel_encoder.measure().expect("wheel encoder faulted"); // Read the encoder value
-		if encoder_value > STOP_THRESHOLD {
+
+		let encoder_value = self.wheel_encoder.lock().unwrap();
+		let distance = encoder_value.get_distance();
+		let velocity: f32 = encoder_value.get_velocity();
```
Comment (taesungh) [inline pod-operation/src/state_machine.rs:260]: Chore: don't think this explicit type annotation is needed

Comment (taesungh) [inline pod-operation/src/state_machine.rs:260]: Thought: unsure if the mutex guard will automatically be dropped or not here. We can try testing by putting a big await time below here and seeing if this prevents the other task from proceeding.

File: pod-operation/src/state_machine.rs [hunk 4]
```diff
@@ -226,21 +248,36 @@ impl StateMachine {
...
+
+		let full_json = json!({
+			"distance": distance,
+			"velocity": velocity,
+		});
```
Comment (taesungh) [inline pod-operation/src/state_machine.rs:265]: Chore: we'll need to nest the values inside another property called `wheel_encoder` as we did before

**Review by taesungh — COMMENTED** [review]
Thanks for working on this. Some notes along with a concern about when mutex becomes unlocked.

**Review by taesungh — APPROVED** [review]
Thanks for adding, note: still need to actually handle the fault so we don't have an incident like last year.

**Pull request merged**